lib/frodata/concerns/api.rb

```diff
@@ -0,0 +1,262 @@
+
+      # Public: Helper methods for performing arbitrary actions against the API using
+      # various HTTP verbs.
+      #
+      # Examples
+      #
+      #   # Perform a get request
+      #   client.get '/api/data/v9.1/leads'
+      #   client.api_get 'leads'
+      #
+      #   # Perform a post request
+      #   client.post '/api/data/v9.1/leads', { ... }
+      #   client.api_post 'leads', { ... }
+      #
+      #   # Perform a put request
+      #   client.put '/api/data/v9.1/leads(073ca9c8-2a41-e911-a81d-000d3a1d5a0b)', { ... }
+      #   client.api_put 'leads(073ca9c8-2a41-e911-a81d-000d3a1d5a0b)', { ... }
+      #
+      #   # Perform a delete request
+      #   client.delete '/api/data/v9.1/leads(073ca9c8-2a41-e911-a81d-000d3a1d5a0b)'
+      #   client.api_delete 'leads(073ca9c8-2a41-e911-a81d-000d3a1d5a0b)'
+      #
+      # Returns the Faraday::Response.
+      define_verbs :get, :post, :put, :delete, :patch, :head
+
+      def metadata
+        api_get("$metadata").body
+      end
+
+      # Public: Executs a SOQL query and returns the result.
+      #
+      # soql - A SOQL expression.
+      #
+      # Examples
+      #
+      #   # Find the names of all Accounts
+      #   client.query('select Name from Account').map(&:Name)
+      #   # => ['Foo Bar Inc.', 'Whizbang Corp']
+      #
+      # Returns a Restforce::Collection if Restforce.configuration.mashify is true.
+      # Returns an Array of Hash for each record in the result if
+      # Restforce.configuration.mashify is false.
+      def query(soql)
+        response = api_get 'query', q: soql
+        mashify? ? response.body : response.body['records']
+      end
+
+      # Public: Insert a new record.
+      #
+      # entity_set - The set the entity belongs to
+      # attrs   - Hash of attributes to set on the new record.
+      #
+      # Examples
+      #
+      #   # Add a new lead
+      #   client.create('leads', {"firstname" =>'Bob'})
+      #   # => '073ca9c8-2a41-e911-a81d-000d3a1d5a0b'
+      #
+      # Returns the String Guid of the newly created entity.
+      # Returns false if something bad happens.
+      def create(*args)
+        create!(*args)
+      rescue *exceptions
+        false
+      end
+      alias insert create
+
+      # Public: Insert a new record.
+      #
+      # entity_set - The set the entity belongs to
+      # attrs   - Hash of attributes to set on the new record.
+      #
+      # Examples
+      #
+      #   # Add a new lead
+      #   client.create!('leads', {"firstname" =>'Bob'})
+      #   # => '073ca9c8-2a41-e911-a81d-000d3a1d5a0b'
+      #
+      # Returns the String Guid of the newly created entity.
+      # Raises exceptions if an error is returned from Dynamics.
+      def create!(entity_set, attrs)
+        entity = service[entity_set].new_entity(attrs)
+        url_chunk = to_url_chunk(entity)
+        url = api_post(url_chunk, attrs).headers['odata-entityid']
+        id = url.match(/\(.+\)/)[0]
```

Comment:
Why are we extracting the id from the url? Why not just get it from the body of the response instead, it will be returned. 

Technically there are places where a key is multiple properties which means the key is a combined value. In the case of multiple keys you can just return a hash of the values?

Reply:
Well maybe this doesn't make sense as this is what the protocol indicates.

```Upon successful completion, the response MUST contain a Location header that contains the edit URL or read URL of the created entity.
Upon successful completion the service MUST respond with either 201 Created, or 204 No Content if the request included a return Prefer header with a value of```

So given that it MUST respond with a URL and it may not respond with the content parsing the url might make sense. Not sure